lib/instrumentation/modules/amqplib.js

```diff
@@ -0,0 +1,54 @@
+'use strict'
+
+var shimmer = require('../shimmer')
+var { getDBDestination } = require('../context')
+
+var URL = require('url-parse')
+
+module.exports = function (amqplib, agent, { version, enabled }) {
+  agent.logger.debug('shimming amqplib.connect')
+  shimmer.wrap(amqplib, 'connect', wrapConnect)
+
+  return amqplib
+
+  function patchChannelModel (channelModel, host, port) {
+    agent.logger.debug('shimming amqplib.ChannelModel.connection.sendMessage')
+    shimmer.wrap(channelModel.connection, 'sendMessage', function (orig) {
+      return function (channel, method, fields, properties, props, content) {
+        var span = enabled && agent.startSpan(null, 'messaging', 'amqp', 'publish')
+        var id = span && span.transaction.id
+
+        agent.logger.debug('intercepted call to amqplib.Connection.sendMessage %o', { id: id, channel: channel, content: content })
+
+        if (span) {
+          span.setDestinationContext(getDBDestination(span, host, port))
+        }
+
+        var res = orig.apply(this, arguments)
+        span.end()
+        return res
+      }
+    })
+  }
+
+  function wrapConnect (original) {
+    return function wrappedConnect (url) {
+      let protocol, host, port
+      if (typeof url === 'object') {
+        protocol = (url.protocol || 'amqp') + ':'
+        host = url.hostname
+        port = url.port || ((protocol === 'amqp:') ? 5672 : 5671)
+      } else {
+        var parts = URL(url, true)
+        protocol = parts.protocol
+        host = parts.hostname
+        port = parseInt(parts.port) || ((protocol === 'amqp:') ? 5672 : 5671)
+      }
+
+      return original.apply(this, arguments).then(function (channelModel) {
+        patchChannelModel(channelModel, host, port)
```

Comment:
Question: I am not too familiar with the amqplib, Does this code handle both the callback and promise version of the API? 

Can we add tests for the same?

Reply:
The callback version is just a wrapping of the promise version so it will work as it.
I can add a test for the callback version if you want.

Reply:
Yeah adding a test should be good here.

Reply:
You were right, some issues with the callback version. I'm digging into.